### PR TITLE
Add validation of trailers to UHV

### PR DESCRIPTION
**Pull request by yanavlasov**
Also consolidate error codes in UHV interface.

Risk Level: Low, not in production build
Testing: Unit Tests
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A

Signed-off-by: Yan Avlasov <yavlasov@google.com>

### CODE REVIEW

File: envoy/http/header_validator.h
Comment (alyssawilk) [inline envoy/http/header_validator.h:89]: should the other functions also include gRPC resposnes or is it trailers specific?

**Review by alyssawilk — COMMENTED** [review]
one nit otherwise LGTM.
/wait

**Review by snowp — DISMISSED** [review]
LGTM after other comments, thanks!

**Comment by wrowe** [comment]
It's my understanding that this shouldn't be staled-out, correct @yanavlasov?

**Comment by alyssawilk** [comment]
Yan what's the state on ths one?
/wait on CI if nothing else

**Comment by yanavlasov** [comment]
> Yan what's the state on ths one? /wait on CI if nothing else

Still WiP. It needs some PRs to come in first that I'm working on.

**Comment by yanavlasov** [comment]
Blocked by #24497

File: envoy/http/header_validator.h
Comment (yanavlasov) [reply envoy/http/header_validator.h:102]: All other functions include gRPC responses if they are possible. Response trailers error can only result on reset.

**Comment by yanavlasov** [comment]
Ready for review now. @wrowe @alyssawilk

File: envoy/http/header_validator.h
Comment (alyssawilk) [inline envoy/http/header_validator.h:101]: nit: should this be HTTP request to be rejected (no s) here and below?

File: source/common/http/conn_manager_impl.cc
Comment (alyssawilk) [inline source/common/http/conn_manager_impl.cc:913]: if (!header_validator_) {
  return true
}
and unindent all below?

File: source/common/http/conn_manager_impl.cc
Comment (alyssawilk) [inline source/common/http/conn_manager_impl.cc:915]: ditto here: if ok() return true?

File: source/common/http/conn_manager_impl.cc
Comment (alyssawilk) [inline source/common/http/conn_manager_impl.cc:934]: why not send a local response for H2 in this case?

File: source/extensions/http/header_validators/envoy_default/header_validator.cc
Comment (alyssawilk) [inline source/extensions/http/header_validators/envoy_default/header_validator.cc:437]: Is this not true for H3 as well?

File: source/extensions/http/header_validators/envoy_default/header_validator.cc
Comment (alyssawilk) [inline source/extensions/http/header_validators/envoy_default/header_validator.cc:438]: I'm missing where we validate pseudo headers?
Also if we don't remove pseudo headers for HTTP/1.1 will we somewhere where we're doing protocol downgrade?

**Review by alyssawilk — COMMENTED** [review]
/wait

File: envoy/http/header_validator.h
Comment (yanavlasov) [reply envoy/http/header_validator.h]: Done

File: source/common/http/conn_manager_impl.cc
Comment (yanavlasov) [reply source/common/http/conn_manager_impl.cc]: Done

File: source/common/http/conn_manager_impl.cc
Comment (yanavlasov) [reply source/common/http/conn_manager_impl.cc]: Done

File: source/common/http/conn_manager_impl.cc
Comment (yanavlasov) [reply source/common/http/conn_manager_impl.cc]: I have filed and Issue and left to TODO to harmonize the behavior. I will do it in a separate PR as it will affect a lot of tests, where the reset for H/2 and H/3 is expected.

File: source/extensions/http/header_validators/envoy_default/header_validator.cc
Comment (yanavlasov) [reply source/extensions/http/header_validators/envoy_default/header_validator.cc]: Updated comment

File: source/extensions/http/header_validators/envoy_default/header_validator.cc
Comment (yanavlasov) [reply source/extensions/http/header_validators/envoy_default/header_validator.cc]: We validate them in [H/1](https://github.com/envoyproxy/envoy/blob/bc3d03886cd4ca8b9677cbaba7976fc2840b9597/source/extensions/http/header_validators/envoy_default/http1_header_validator.cc#L53) and [H/2](https://github.com/envoyproxy/envoy/blob/bc3d03886cd4ca8b9677cbaba7976fc2840b9597/source/extensions/http/header_validators/envoy_default/http2_header_validator.cc#L50) specific validators. Although the function is the same in both cases and should be refactored (I'll file an Issue for this). The `validateGenericHeaderName` fails if there is ':' in the name.

For H/1 we never remove pseudoheaders, since the H/1 parser can not produce them. We only add them for the request and response using the first line from the wire. And we never add them for trailers.

Comment (alyssawilk) [reply source/extensions/http/header_validators/envoy_default/header_validator.cc]: ah, the comment reads like pseudo headers are allowed for HTTP/1.1   maybe make that clear?

File: envoy/http/header_validator.h
Comment (alyssawilk) [inline envoy/http/header_validator.h:80]: sorry my correction needs a correction :-(
I was thinking "the HTTP request"
"http requests" was also fine

File: source/common/http/conn_manager_impl.h
Comment (alyssawilk) [inline source/common/http/conn_manager_impl.h:357]: comment?

File: source/extensions/http/header_validators/envoy_default/header_validator.cc
Comment (alyssawilk) [inline source/extensions/http/header_validators/envoy_default/header_validator.cc:441]: for H2 and H3 no pseudo headerrs.  if we disallow for H1 as well let's say there's no prohibitions but we disallow anyway?

**Review by alyssawilk — COMMENTED** [review]
LG modulo a few remaining nits

File: envoy/http/header_validator.h
Comment (yanavlasov) [reply envoy/http/header_validator.h]: Changed to "the HTTP request".

File: source/common/http/conn_manager_impl.h
Comment (yanavlasov) [reply source/common/http/conn_manager_impl.h:366]: Done

File: source/extensions/http/header_validators/envoy_default/header_validator.cc
Comment (yanavlasov) [reply source/extensions/http/header_validators/envoy_default/header_validator.cc]: Yes, the comment is confusing. I've tried to clarify it, see if it is better. I'd like to provide more details from RFCs on why the common function would work.

File: source/extensions/http/header_validators/envoy_default/header_validator.cc
Comment (yanavlasov) [reply source/extensions/http/header_validators/envoy_default/header_validator.cc]: Clarified comment. I can also remove the more detailed explanation and just link to RFCs.

**Review by alyssawilk — APPROVED** [review]

**Pull request merged**
